From 092022bcc5255e90eab11909021fa631d3944e79 Mon Sep 17 00:00:00 2001
From: Corey Bryant <corey.bryant@canonical.com>
Date: Fri, 27 Jan 2023 17:02:53 -0500
Subject: [PATCH 09/86] Revert "Bug#34824515 concurrent_map::for_each does not
 support pass-by-value"

This reverts commit 729783fd3bcc378bd6f027383b9b30ea40c9890d.
---
 router/src/routing/src/connection_container.cc | 9 +++++++--
 router/src/routing/src/connection_container.h  | 2 +-
 2 files changed, 8 insertions(+), 3 deletions(-)

diff --git a/router/src/routing/src/connection_container.cc b/router/src/routing/src/connection_container.cc
index fd59fd6e29c..fc410860ca4 100644
--- a/router/src/routing/src/connection_container.cc
+++ b/router/src/routing/src/connection_container.cc
@@ -60,8 +60,13 @@ unsigned ConnectionContainer::disconnect(const AllowedNodes &nodes) {
 }
 
 void ConnectionContainer::disconnect_all() {
-  connections_.for_each(
-      [](auto &connection) { connection.first->disconnect(); });
+  auto mark_to_disconnect =
+      [](std::pair<MySQLRoutingConnectionBase *const,
+                   std::unique_ptr<MySQLRoutingConnectionBase>> &connection) {
+        connection.first->disconnect();
+      };
+
+  connections_.for_each(mark_to_disconnect);
 }
 
 void ConnectionContainer::remove_connection(
diff --git a/router/src/routing/src/connection_container.h b/router/src/routing/src/connection_container.h
index 84c71e2865c..5cfad3d2c88 100644
--- a/router/src/routing/src/connection_container.h
+++ b/router/src/routing/src/connection_container.h
@@ -69,7 +69,7 @@ class concurrent_map {
   }
 
   template <typename Predicate>
-  void for_each(Predicate p) {
+  void for_each(Predicate &p) {
     for (auto &each_bucket : buckets_) {
       each_bucket.for_each(p);
     }
-- 
2.37.2

